src/dtx/dtx_common.c

```diff
@@ -663,3 +653,205 @@ dtx_handle_resend(daos_handle_t coh, daos_unit_oid_t *oid,
+		struct dtx_conflict_entry	*conflict;
+		int				shard_cnt = exec_arg->shard_cnt;
+		int				i;
+		int				j;
+
+		D_ALLOC_ARRAY(conflict, shard_cnt);
+		if (conflict == NULL) {
+			rc = -DER_NOMEM;
+			goto out;
+		}
+
+		shard_arg = (struct dtx_exec_shard_arg *)(exec_arg + 1);
+		for (i = 0, j = 0; i < shard_cnt; i++, shard_arg++) {
+			if (!daos_is_zero_dti(&shard_arg->exec_dce.dce_xid)) {
+				daos_dti_copy(&conflict[j].dce_xid,
+					      &shard_arg->exec_dce.dce_xid);
+				conflict[j++].dce_dkey =
+						shard_arg->exec_dce.dce_dkey;
+			}
+		}
+
+		D_ASSERT(j > 0);
+
+		*dces = conflict;
+		rc = j;
+	}
+
+out:
+	dtx_exec_arg_free(exec_arg);
+
+	return rc;
+};
+
+void
+dtx_exec_ops_ult(void *arg)
+{
+	struct dtx_exec_arg	*exec_arg = arg;
+	ABT_future		future = exec_arg->future;
+	struct dtx_exec_shard_arg *shard_arg;
+	uint32_t		i;
+	int			shard_cnt = exec_arg->shard_cnt;
+	int			rc = 0;
+
+	D_ASSERT(shard_cnt >= 1);
+	D_ASSERT(future != ABT_FUTURE_NULL);
+	shard_arg = (struct dtx_exec_shard_arg *)(exec_arg + 1);
+	for (i = 0; i < shard_cnt; i++, shard_arg++) {
+		rc = dtx_exec_op(shard_arg, i);
+		if (rc != 0)
+			break;
+	}
+
+	if (rc != 0) {
+		D_ASSERT(i < shard_cnt);
+		for (i++; i < shard_cnt; i++) {
+			shard_arg++;
+			shard_arg->exec_shard_rc = rc;
+			rc = ABT_future_set(future, shard_arg);
+			D_ASSERTF(rc == ABT_SUCCESS,
+				  "ABT_future_set failed %d.\n", rc);
+		}
+	}
+}
+
+/**
+ * Execute the operations on all targets.
+ */
+int
+dtx_exec_ops(struct daos_shard_tgt *shard_tgts, int tgts_cnt,
+	     struct dtx_handle *dth, dtx_exec_shard_func_t exec_func,
+	     void *func_arg)
+{
+	struct dtx_exec_arg		*exec_arg;
+	struct dtx_exec_shard_arg	*shard_arg;
+	ABT_future			future;
+	int				i;
+	int				rc;
+
+	D_ALLOC(exec_arg, sizeof(struct dtx_exec_arg) +
+		tgts_cnt * sizeof(struct dtx_exec_shard_arg));
+	if (exec_arg == NULL)
+		return -DER_NOMEM;
+
+	rc = ABT_future_create(tgts_cnt, dtx_exec_ops_comp_cb, &future);
+	if (rc != ABT_SUCCESS) {
+		D_ERROR("ABT_future_create failed %d.\n", rc);
+		D_FREE(exec_arg);
+		return dss_abterr2der(rc);
+	}
+
+	exec_arg->exec_func	= exec_func;
+	exec_arg->exec_func_arg	= func_arg;
+	exec_arg->future	= future;
+	exec_arg->shard_cnt	= tgts_cnt;
+	exec_arg->dth		= dth;
+	shard_arg = (struct dtx_exec_shard_arg *)(exec_arg + 1);
+	for (i = 0; i < tgts_cnt; i++, shard_arg++) {
+		shard_arg->exec_shard_tgt = shard_tgts + i;
+		shard_arg->exec_arg	= exec_arg;
+	}
+
+	dth->dth_exec_arg = exec_arg;
+
+	/*
+	 * XXX ideally, we probably should create ULT for each shard, but
+	 * for performance reasons, let's create one for all remote targets
+	 * for now.
+	 */
+	rc = dss_ult_create(dtx_exec_ops_ult, exec_arg, DSS_ULT_IOFW,
+			    dss_get_module_info()->dmi_tgt_id, 0, NULL);
+	if (rc != 0) {
+		D_ERROR("ult create failed %d.\n", rc);
+		dtx_exec_arg_free(exec_arg);
+		D_GOTO(out, rc);
+	}
+
+	/* Then execute the local operation */
+	rc = exec_func(dth, func_arg, -1, NULL, NULL);
+out:
+	return rc;
+}
```

Comment:
What's the point of introducing these RPC dispatch functions in DTX? Shouldn't we put them in your newly added srv_obj_remote.c?

Reply:
Hmm, not sure I follow here. This is not RPC dispatch, this is to tell all parties to execute its own part of the distribute transaction.  If it is local, call vos, otherwise call obj_remote_xxx in src_obj_remote.c

Reply:
Let's say in this way, what's the point of calling the RPC dispatch (obj_remote_xxx) in a callback in DTX, why don't we just directly call them in obj module?